### PR TITLE
Add pagination / document limit logic for reranking

### PR DESCRIPTION
**Pull request by msfroh**
### Description
This change augments https://github.com/opensearch-project/search-relevance/pull/26 by adding unit tests.

* Add `preprocessRequest` interface to ResultTransformer.
* Override `from` and `size` parameters during preprocess, to fetch more than the default 10 hits from OpenSearch (improved accuracy if the ranker can access more documents)
* Limit the number of documents sent to Kendra Ranking (Will make this number configurable in a follow up)
* Populate response documents from re-ranked hits (upto Kendra limit) and OpenSearch hits beyond the limit

### Issues Resolved
N/A
 
### Check List
- [X] New functionality includes testing.
  - [X] All tests pass
- [X] New functionality has been documented.
  - [X] New functionality has javadoc added
- [X] Commits are signed as per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/search/relevance/transformer/kendraintelligentranking/KendraIntelligentRanker.java [hunk 1]
```diff
@@ -126,6 +127,11 @@ public SearchHits transform(final SearchHits hits,
         kendraConfig.getProperties().getBodyFields(),
         kendraConfig.getProperties().getTitleFields());
     if (queryParserResult == null) {
+      // Unknown query type or query does not reference body field
+      return hits;
+    }
+    if (hits.getHits().length == 0) {
```
Comment (mahitamahesh) [inline src/main/java/org/opensearch/search/relevance/transformer/kendraintelligentranking/KendraIntelligentRanker.java]: Nit: If we check this right up front, we can avoid the query parser step too

File: src/test/java/org/opensearch/search/relevance/transformer/kendraintelligentranking/KendraIntelligentRankerTests.java [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+    public void testShouldNotTransformWithoutSource() {
+        KendraIntelligentRanker ranker = new KendraIntelligentRanker(buildMockHttpClient());
+        SearchRequest originalRequest = new SearchRequest();
+        boolean shouldTransform = ranker.shouldTransform(originalRequest, new KendraIntelligentRankingConfiguration());
+        assertFalse(shouldTransform);
+    }
+
+    public void testShouldNotTransformWithoutQuery() {
+        KendraIntelligentRanker ranker = new KendraIntelligentRanker(buildMockHttpClient());
+        SearchRequest originalRequest = new SearchRequest();
```
Comment (mahitamahesh) [inline src/test/java/org/opensearch/search/relevance/transformer/kendraintelligentranking/KendraIntelligentRankerTests.java]: We should set the SearchSourceBuilder with query(null) here

Comment (msfroh) [reply src/test/java/org/opensearch/search/relevance/transformer/kendraintelligentranking/KendraIntelligentRankerTests.java]: Oops! Thanks

I think I copy/pasted the test above, got distracted, then came back and started writing the next test. 😁

File: src/main/java/org/opensearch/search/relevance/transformer/kendraintelligentranking/KendraIntelligentRanker.java (hunk 1 above)
Comment (msfroh) [reply src/main/java/org/opensearch/search/relevance/transformer/kendraintelligentranking/KendraIntelligentRanker.java]: Good catch, thanks!

**Review by mahitamahesh — APPROVED** [review]

**Comment by msfroh** [comment]
Yay for randomized testing! 

My earlier commit failed on the case where `testTransformHits` randomly selected zero as a docLimit and zero additional hits. That's already covered by the earlier `testTransformHits` test case so we can safely skip it. 

Still, it's nice to know that we're considering many possibilities.

**Review by macohen — APPROVED** [review]

**Pull request merged**